<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Converting single-cell data structures between Bioconductor and Python â€¢ zellkonverter</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Converting single-cell data structures between Bioconductor and Python">
<meta property="og:description" content="zellkonverter">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">zellkonverter</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.5.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/zellkonverter.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/theislab/zellkonverter/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Converting single-cell data structures between Bioconductor and Python</h1>
                        <h4 data-toc-skip class="author">Luke Zappia</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:luke@lazappi.id.au" class="email">luke@lazappi.id.au</a>
      
                              <h4 data-toc-skip class="author">Aaron Lun</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:infinite.monkeys.with.keyboards@gmail.com" class="email">infinite.monkeys.with.keyboards@gmail.com</a>
      
                  
            <h4 data-toc-skip class="date">Revised: 19 October 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/theislab/zellkonverter/blob/HEAD/vignettes/zellkonverter.Rmd" class="external-link"><code>vignettes/zellkonverter.Rmd</code></a></small>
      <div class="hidden name"><code>zellkonverter.Rmd</code></div>

    </div>

    
    
<div class="section level1">
<h1 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h1>
<p>This package provides a lightweight interface between the Bioconductor <code>SingleCellExperiment</code> data structure and the Python <code>AnnData</code>-based single-cell analysis environment. The idea is to enable users and developers to easily move data between these frameworks to construct a multi-language analysis pipeline across R/Bioconductor and Python.</p>
</div>
<div class="section level1">
<h1 id="reading-and-writing-h5ad-files">Reading and writing H5AD files<a class="anchor" aria-label="anchor" href="#reading-and-writing-h5ad-files"></a>
</h1>
<p>The <code><a href="../reference/readH5AD.html">readH5AD()</a></code> function can be used to read a <code>SingleCellExperiment</code> from a H5AD file. This can be manipulated in the usual way as described in the <em><a href="https://bioconductor.org/packages/3.15/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em> documentation.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/theislab/zellkonverter" class="external-link">zellkonverter</a></span><span class="op">)</span>

<span class="co"># Obtaining an example H5AD file.</span>
<span class="va">example_h5ad</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"krumsiek11.h5ad"</span>,
                            package <span class="op">=</span> <span class="st">"zellkonverter"</span><span class="op">)</span>
<span class="fu"><a href="../reference/readH5AD.html">readH5AD</a></span><span class="op">(</span><span class="va">example_h5ad</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## class: SingleCellExperiment </span>
<span class="co">## dim: 11 640 </span>
<span class="co">## metadata(2): highlights iroot</span>
<span class="co">## assays(1): X</span>
<span class="co">## rownames(11): Gata2 Gata1 ... EgrNab Gfi1</span>
<span class="co">## rowData names(0):</span>
<span class="co">## colnames(640): 0 1 ... 158-3 159-3</span>
<span class="co">## colData names(1): cell_type</span>
<span class="co">## reducedDimNames(0):</span>
<span class="co">## mainExpName: NULL</span>
<span class="co">## altExpNames(0):</span></code></pre>
<p>We can also write a <code>SingleCellExperiment</code> to a H5AD file with the <code><a href="../reference/writeH5AD.html">writeH5AD()</a></code> function. This is demonstrated below on the classic Zeisel mouse brain dataset from the <em><a href="https://bioconductor.org/packages/3.15/scRNAseq" class="external-link">scRNAseq</a></em> package. The resulting file can then be directly used in compatible Python-based analysis frameworks.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span>

<span class="va">sce_zeisel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/ZeiselBrainData.html" class="external-link">ZeiselBrainData</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">out_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".h5ad"</span><span class="op">)</span>
<span class="fu"><a href="../reference/writeH5AD.html">writeH5AD</a></span><span class="op">(</span><span class="va">sce_zeisel</span>, file <span class="op">=</span> <span class="va">out_path</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="converting-between-singlecellexperiment-and-anndata-objects">Converting between <code>SingleCellExperiment</code> and <code>AnnData</code> objects<a class="anchor" aria-label="anchor" href="#converting-between-singlecellexperiment-and-anndata-objects"></a>
</h1>
<p>Developers and power users who control their Python environments can directly convert between <code>SingleCellExperiment</code> and <a href="https://anndata.readthedocs.io/en/stable/" class="external-link"><code>AnnData</code> objects</a> using the <code><a href="../reference/AnnData-Conversion.html">SCE2AnnData()</a></code> and <code><a href="../reference/AnnData-Conversion.html">AnnData2SCE()</a></code> utilities. These functions expect that <em><a href="https://CRAN.R-project.org/package=reticulate" class="external-link">reticulate</a></em> has already been loaded along with an appropriate version of the <a href="https://pypi.org/project/anndata/" class="external-link"><em>anndata</em></a> package. We suggest using the <em><a href="https://bioconductor.org/packages/3.15/basilisk" class="external-link">basilisk</a></em> package to set up the Python environment before using these functions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">basilisk</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span>

<span class="va">seger</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/SegerstolpePancreasData.html" class="external-link">SegerstolpePancreasData</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">roundtrip</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/basilisk/man/basiliskStart.html" class="external-link">basiliskRun</a></span><span class="op">(</span>fun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">{</span>
     <span class="co"># Convert SCE to AnnData:</span>
     <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AnnData-Conversion.html">SCE2AnnData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>

     <span class="co"># Maybe do some work in Python on 'adata':</span>
     <span class="co"># BLAH BLAH BLAH</span>

     <span class="co"># Convert back to an SCE:</span>
     <span class="fu"><a href="../reference/AnnData-Conversion.html">AnnData2SCE</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span>
<span class="op">}</span>, env <span class="op">=</span> <span class="va">zellkonverterAnnDataEnv</span>, sce <span class="op">=</span> <span class="va">seger</span><span class="op">)</span></code></pre></div>
<p>Package developers can guarantee that they are using the same versions of Python packages as <em><a href="https://bioconductor.org/packages/3.15/zellkonverter" class="external-link">zellkonverter</a></em> by using the <code>.AnnDataDependencies</code> variable to set up their Python environments.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">.AnnDataDependencies</span></code></pre></div>
<pre><code><span class="co">## [1] "anndata==0.7.6"  "h5py==3.2.1"     "hdf5==1.10.6"    "natsort==7.1.1" </span>
<span class="co">## [5] "numpy==1.20.2"   "packaging==20.9" "pandas==1.2.4"   "scipy==1.6.3"   </span>
<span class="co">## [9] "sqlite==3.35.5"</span></code></pre>
</div>
<div class="section level1">
<h1 id="progress-messages">Progress messages<a class="anchor" aria-label="anchor" href="#progress-messages"></a>
</h1>
<p>By default the functions in <em><a href="https://bioconductor.org/packages/3.15/zellkonverter" class="external-link">zellkonverter</a></em> donâ€™t display any information about their progress but this can be turned on by setting the <code>verbose = TRUE</code> argument.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/readH5AD.html">readH5AD</a></span><span class="op">(</span><span class="va">example_h5ad</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## class: SingleCellExperiment </span>
<span class="co">## dim: 11 640 </span>
<span class="co">## metadata(2): highlights iroot</span>
<span class="co">## assays(1): X</span>
<span class="co">## rownames(11): Gata2 Gata1 ... EgrNab Gfi1</span>
<span class="co">## rowData names(0):</span>
<span class="co">## colnames(640): 0 1 ... 158-3 159-3</span>
<span class="co">## colData names(1): cell_type</span>
<span class="co">## reducedDimNames(0):</span>
<span class="co">## mainExpName: NULL</span>
<span class="co">## altExpNames(0):</span></code></pre>
<p>If you would like to see progress messages for all functions by default you can turn this on using the <code><a href="../reference/setZellkonverterVerbose.html">setZellkonverterVerbose()</a></code> function.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># This is not run here </span>
<span class="fu"><a href="../reference/setZellkonverterVerbose.html">setZellkonverterVerbose</a></span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h1>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R Under development (unstable) (2021-10-28 r81109)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.3 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats    graphics utils    stats4   methods  base    </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] basilisk_1.7.0              scRNAseq_2.9.0             </span>
<span class="co">##  [3] SingleCellExperiment_1.17.0 SummarizedExperiment_1.25.1</span>
<span class="co">##  [5] Biobase_2.55.0              GenomicRanges_1.47.3       </span>
<span class="co">##  [7] GenomeInfoDb_1.31.1         IRanges_2.29.0             </span>
<span class="co">##  [9] S4Vectors_0.33.0            BiocGenerics_0.41.1        </span>
<span class="co">## [11] MatrixGenerics_1.7.0        matrixStats_0.61.0         </span>
<span class="co">## [13] zellkonverter_1.5.0         knitr_1.36                 </span>
<span class="co">## [15] BiocStyle_2.23.0           </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] rjson_0.2.20                  ellipsis_0.3.2               </span>
<span class="co">##   [3] rprojroot_2.0.2               XVector_0.35.0               </span>
<span class="co">##   [5] fs_1.5.0                      bit64_4.0.5                  </span>
<span class="co">##   [7] interactiveDisplayBase_1.33.0 AnnotationDbi_1.57.1         </span>
<span class="co">##   [9] fansi_0.5.0                   xml2_1.3.2                   </span>
<span class="co">##  [11] cachem_1.0.6                  jsonlite_1.7.2               </span>
<span class="co">##  [13] Rsamtools_2.11.0              dbplyr_2.1.1                 </span>
<span class="co">##  [15] png_0.1-7                     shiny_1.7.1                  </span>
<span class="co">##  [17] BiocManager_1.30.16           compiler_4.2.0               </span>
<span class="co">##  [19] httr_1.4.2                    lazyeval_0.2.2               </span>
<span class="co">##  [21] assertthat_0.2.1              Matrix_1.3-4                 </span>
<span class="co">##  [23] fastmap_1.1.0                 cli_3.1.0                    </span>
<span class="co">##  [25] later_1.3.0                   htmltools_0.5.2              </span>
<span class="co">##  [27] prettyunits_1.1.1             tools_4.2.0                  </span>
<span class="co">##  [29] glue_1.4.2                    GenomeInfoDbData_1.2.7       </span>
<span class="co">##  [31] dplyr_1.0.7                   grDevices_4.2.0              </span>
<span class="co">##  [33] rappdirs_0.3.3                Rcpp_1.0.7                   </span>
<span class="co">##  [35] jquerylib_0.1.4               pkgdown_1.9000.9000.9000     </span>
<span class="co">##  [37] vctrs_0.3.8                   Biostrings_2.63.0            </span>
<span class="co">##  [39] ExperimentHub_2.3.0           rtracklayer_1.55.0           </span>
<span class="co">##  [41] xfun_0.27                     stringr_1.4.0                </span>
<span class="co">##  [43] mime_0.12                     lifecycle_1.0.1              </span>
<span class="co">##  [45] restfulr_0.0.13               ensembldb_2.19.0             </span>
<span class="co">##  [47] XML_3.99-0.8                  AnnotationHub_3.3.0          </span>
<span class="co">##  [49] zlibbioc_1.41.0               basilisk.utils_1.7.0         </span>
<span class="co">##  [51] ProtGenerics_1.27.0           ragg_1.2.0                   </span>
<span class="co">##  [53] hms_1.1.1                     promises_1.2.0.1             </span>
<span class="co">##  [55] parallel_4.2.0                AnnotationFilter_1.19.0      </span>
<span class="co">##  [57] yaml_2.2.1                    curl_4.3.2                   </span>
<span class="co">##  [59] memoise_2.0.0                 reticulate_1.22              </span>
<span class="co">##  [61] sass_0.4.0                    datasets_4.2.0               </span>
<span class="co">##  [63] biomaRt_2.51.0                stringi_1.7.5                </span>
<span class="co">##  [65] RSQLite_2.2.8                 BiocVersion_3.15.0           </span>
<span class="co">##  [67] BiocIO_1.5.0                  desc_1.4.0                   </span>
<span class="co">##  [69] GenomicFeatures_1.47.2        filelock_1.0.2               </span>
<span class="co">##  [71] BiocParallel_1.29.0           rlang_0.4.12                 </span>
<span class="co">##  [73] pkgconfig_2.0.3               systemfonts_1.0.3            </span>
<span class="co">##  [75] bitops_1.0-7                  evaluate_0.14                </span>
<span class="co">##  [77] lattice_0.20-45               purrr_0.3.4                  </span>
<span class="co">##  [79] GenomicAlignments_1.31.0      bit_4.0.4                    </span>
<span class="co">##  [81] tidyselect_1.1.1              here_1.0.1                   </span>
<span class="co">##  [83] magrittr_2.0.1                bookdown_0.24                </span>
<span class="co">##  [85] R6_2.5.1                      generics_0.1.1               </span>
<span class="co">##  [87] DelayedArray_0.21.1           DBI_1.1.1                    </span>
<span class="co">##  [89] withr_2.4.2                   pillar_1.6.4                 </span>
<span class="co">##  [91] KEGGREST_1.35.0               RCurl_1.98-1.5               </span>
<span class="co">##  [93] tibble_3.1.5                  dir.expiry_1.3.0             </span>
<span class="co">##  [95] crayon_1.4.2                  utf8_1.2.2                   </span>
<span class="co">##  [97] BiocFileCache_2.3.0           rmarkdown_2.11               </span>
<span class="co">##  [99] progress_1.2.2                grid_4.2.0                   </span>
<span class="co">## [101] blob_1.2.2                    digest_0.6.28                </span>
<span class="co">## [103] xtable_1.8-4                  httpuv_1.6.3                 </span>
<span class="co">## [105] textshaping_0.3.6             bslib_0.3.1</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Luke Zappia, Aaron Lun.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
